### PR TITLE
Test payload snapshots

### PR DESCRIPTION
**Pull request by gnode1**
Adds node snapshot tests both on local machine and on CI.

Add java snapshot on local machine only. Java has no real tests on CI because it uses bats. Need to enable that in java directory.

Discovered bug in recovery of open sockets. We populate our filesys structure BUT we don't open file on the host. Next open file on the host will conflict with the our data. The test for this is Java snapshot test which I only ran locally (see above).

I'd really like to push this ASAP and figure out CI in Java after that.

### CODE REVIEW

**Review by sv641 — APPROVED** [review]

File: km/km_filesys.c [hunk 1]
```diff
@@ -2660,13 +2650,13 @@ static int km_fs_recover_open_file(char* ptr, size_t length)
 
    int fd = open(name, nt_file->flags, 0);
```
Comment (paulpopelka) [inline km/km_filesys.c:2651]: codeql has decided we read a filename from user input here and then opening the file.

It seems like codeql has not gone deep enough to discover what km has opened and is reading the filename from.

We are reading from a snapshot file.  I guess codeql thinks that users manually generate snapshot files?

I guess we need to investigate what codeql is having fits about.

File: km/km_filesys.c [hunk 2]
```diff
@@ -2537,7 +2537,13 @@ static inline int km_fs_recover_socket(km_nt_socket_t* nt_sock, struct sockaddr*
       return -1;
    }
    km_fs_recover_fd(nt_sock->fd, host_fd, 0, km_get_nonfile_name(host_fd), -1, nt_sock->how);
-
+   if (nt_sock->state == KM_NT_SKSTATE_ERROR) {
+      file->error = -ECONNRESET;
+      if (file->name != NULL) {
+         free(file->name);
+      }
+      file->name = strdup("socket error");
```
Comment (paulpopelka) [inline km/km_filesys.c:2545]: Check for allocation failure.

File: km/km_filesys.c [hunk 3]
```diff
@@ -2537,7 +2537,13 @@ static inline int km_fs_recover_socket(km_nt_socket_t* nt_sock, struct sockaddr*
       return -1;
    }
    km_fs_recover_fd(nt_sock->fd, host_fd, 0, km_get_nonfile_name(host_fd), -1, nt_sock->how);
-
+   if (nt_sock->state == KM_NT_SKSTATE_ERROR) {
+      file->error = -ECONNRESET;
+      if (file->name != NULL) {
```
Comment (paulpopelka) [inline km/km_filesys.c:2542]: free() tolerates being passed a null pointer.  The man page even says so.

PS: this usually causes endless hours of developers having religious discussions over checking for null before calling free() or letting free() handle it.

File: km/km_filesys.c (hunk 1 above)
Comment (gnode1) [reply km/km_filesys.c:2651]: We can squelch this, tell it it's false positive

File: km/km_filesys.c (hunk 3 above)
Comment (gnode1) [reply km/km_filesys.c:2542]: We have a number of places when we do it this way, and the other way. Should we go over the whole things and make it uniform?

**Pull request merged**